Support handling incoming calls for HFP + multi-SIM

alfonsosanchezbeato opened the pull request:
Needed for blueZ 5.

morphis commented on src/emulator.c:
```diff
@@ -1326,6 +1520,12 @@ void ofono_emulator_set_indicator(struct ofono_emulator *em,
+	if (!valid_indication(em, atom, name))
+		return;
+
+	ofono_info("BTOOTH %s\t%d", name, value);
```
Should be a DBG and needs to be a bit more expressive.

morphis commented on src/emulator.c:
```diff
@@ -1326,6 +1520,12 @@ void ofono_emulator_set_indicator(struct ofono_emulator *em,
 	gboolean waiting;
+	struct ofono_emulator *em = __ofono_atom_get_data(atom);
+
+	if (!valid_indication(em, atom, name))
```
If someone calls ofono_emulator_set_indicator he already knows that the modem supports the OFONO_ATOM_TYPE_EMULATOR_HFP atom and we don't have to check that again. See src/network.c line 1380

alfonsosanchezbeato replied:
valid_indication() filters out indicators that are not from the currently preferred modem. The HF can only display indicators from one modem, so it would be confusing to see, say, signals from 2 different operators so the signal strength indicator flickers all the time.

morphis replied:
I know. However I don't like us passing the atom in ofono_emulator_set_indicator as the name indicates it operates on a struct ofono_emulator rather than a ofono_atom.

alfonsosanchezbeato replied:
I would have simply added the "struct ofono_atom" as a function parameter instead of replacing "struct ofono_emulator", but seemed to me senseless as you can obtain the emulator from the atom.

Changing the name to ofono_atom_*() does not really make sense either.

To really fix this we would need a real implementation of the hfp atom (src/hfp.c?) instead of it being a dummy atom that points to the emulator. Each modem would have an instance of "struct ofono_hfp", and the instances would point to a the single emulator instance. But that seemed to me a too big change.

alfonsosanchezbeato commented:
Refreshed after rebasing from latest main. Comments addressed.

tonyespy commented:
OK, I have a few inline comments, and then some general commentary here.

First, in the future please to try to include an overview of the why the change is being made.  As there's no bug, I had to dig back to the asana task which just states that incoming HFP calls only support a single modem with Bluez5, and that this is regression over Bluez4 ( where HFP was handled outside ofono ).  What's the limitation that causes the issue... the fact that there's only one emulator created for a single modem now?

Also, just to confirm, the DUN behavior still only supports a single SIM.  Why was the selection of the SIM for DUN changed from first to last?

Second, as this is a major API change to ofono_emulator, I wonder how this ought to be handled?  Is there precedent for this in the project?

As with Simon, the usage of ofono_atom * as a parameter in the change caught my eye, but not for the exact same reason.  There are other ofono functions that don't take the primary struct associated with the atom ( eg. ofono_sim_add_file_watch ).  My only hesitation is that ofono_atom isn't exposed in any of the public ofono headers.  Would it be possible to use a modem instance, and then lookup the atom using the modem in the emulator code?

What happens if more than one HFP device attaches?  This looks like it'll just go ahead an create a 2nd HFP emulator instance...

tonyespy commented on src/emulator.c:
```diff
@@ -1506,8 +1670,38 @@ enum ofono_emulator_request_type ofono_emulator_request_get_type(
+
+	/* Return FALSE if the modem is not the preferred one */
+	/*
+	 * TODO For the moment we just take the firs on the list, we need to ask
```
typo: "we just the *first* on the list"

tonyespy commented on examples/emulator.c:
```diff
@@ -63,11 +64,12 @@ static gboolean on_socket_connected(GIOChannel *chan, GIOCondition cond,
-	/* Pick the first powered modem */
-	modem = modems->data;
+	/* Pick the last powered modem */
+	last = g_list_last(modems);
```
g_list_last can return *NULL*

alfonsosanchezbeato replied:
Not really, powered_watch() and modem_watch() make sure there is a modem before this function is ever called. And also remove the watch when the modem list is empty. Also, there was no check in the previous code.

tonyespy commented on plugins/dun_gw_bluez4.c:
```diff
@@ -90,11 +91,12 @@ static void dun_gw_connect_cb(GIOChannel *io, GError *err, gpointer user_data)
-	/* Pick the first powered modem */
-	modem = modems->data;
+	/* Pick the last powered modem */
+	last = g_list_last(modems);
```
g_list_last can return *NULL*

alfonsosanchezbeato replied:
That can happen for the same reasons as stated in the previous comment (see gprs_watch()).

tonyespy commented on plugins/dun_gw_bluez5.c:
```diff
@@ -88,10 +89,11 @@ static DBusMessage *profile_new_connection(DBusConnection *conn,
 	}
 
-	modem = modems->data;
+	last = g_list_last(modems);
```
g_list_last can return *NULL*

alfonsosanchezbeato replied:
There is actually a check in this case for an empty list before this call.

tonyespy commented on plugins/hfp_ag_bluez4.c:
```diff
@@ -99,14 +100,15 @@ static void hfp_ag_connect_cb(GIOChannel *io, GError *err, gpointer user_data)
-	/* Pick the first voicecall capable modem */
-	modem = modems->data;
+	/* Pick the last voicecall capable modem */
+	last = g_list_last(modems);
```
g_list_last can return *NULL*

alfonsosanchezbeato replied:
That cannot happen for the same reasons as stated for examples/emulator.c (see sim_state_watch()). Also, there is a senseless check right after this for a NULL modem, which seems wrong to me.

alfonsosanchezbeato commented:
Thanks for the review. Answering your comments:

* Sure, I will try to be more explicit about the purpose of a PR in the future
* The limitation is not that there is one emulator per modem. The emulator is more attached to the HFP headset than to the modems, so there must be one per headset. The limitation is that the emulator was not able to handle more than one modem, which is what this PR fixes.
* Changing the selection to the last modem in the plugins that still allow you to use only one modem was for a practical reason: in this way there is no need to create a new GList containing only one modem. Instead, you pass a pointer to the last node, and ofono_emulator_create() will add only that modem to the list of modems handled by the emulator. But I have just noted that the list passed to the emulator is a GList instead of a GSList, which kind of makes this approach a bit less elegant. I will add comments in any case.
* Not sure there is a precedent for this kind of changes
* About using ofono_modem instead of ofono_atom, that is a good point. As a minimum it will introduce a bit more of type safety (currently you could pass to the emulator an atom that is not of HFP type), so I will change that. Note that, as I have already stated in the discussion with Simon, the ideal solution from my POV would be to have a real hfp atom that would contain a pointer to the emulator instead having that pointer as a the payload in the atom. In this way we would have one atom per modem, and an emulator shared by all HFP atoms. However, I think that will deviate our code too much from upstream so I do not think it is worth the effort.
* Not sure what happens if more than one HFP device registers in BlueZ, but note that if that creates another emulator it will be the same behaviour as without this PR.
* Ok, I can try to upstream. It will be rejected, but maybe they have interesting comments.

alfonsosanchezbeato commented:
Branch refreshed after addressing comments. However, after having thought about it for a while it does not seem feasible to change the interface to use ofono_modem instead of ofono_atom, so I have not changed that. Reasons are:

1. We can have emulator instances to handle HFP and DUN profiles, so it is possible to have more than one emulator in the system. We could work around that by using ofono_modem + emulator type, but:
2. It is also possible to have more than one HFP connected to the phone, each one with its own emulator instance. The atoms already take this into account, as they use __ofono_modem_foreach_atom(TYPE_EMULATOR_HFP) for calls to the emulator. I can see this should work for controlling calls from two headsets (accept, dial, etc.). Admittedly, I have no idea of what would happen with the audio in that case though, but we should try to keep things generic so this is a supported scenario in the future.

tonyespy commented:
Approved
